rubin_scheduler/site_models/almanac.py

```diff
@@ -80,18 +81,97 @@ def get_planet_positions(self, mjd):
+                ('sun_n12_setting', '<f8'),
+                ('sun_n18_setting', '<f8'),
+                ('sun_n18_rising', '<f8'),
+                ('sun_n12_rising', '<f8'),
+                ('sunrise', '<f8'),
+                ('moonrise', '<f8'),
+                ('moonset', '<f8')
+            ])
         """
-        Returns a numpy array with mjds for various events (sunset, moonrise, sun at -12 degrees alt, etc.).
-        Also the integer night number.
-        """
-        indx = np.searchsorted(self.sunsets["sunset"], mjd, side="right") - 1
+        if mjd is not None and evening_date is not None:
+            raise ValueError("At most one of mjd and evening_date can be set")
+
+        # Default to now
+        if mjd is None and evening_date is None:
+            mjd = 40587 + datetime.datetime.now().timestamp() / (24 * 60 * 60)
+
+        if mjd is not None:
+            indx = np.searchsorted(self.sunsets["sunset"], mjd, side="right") - 1
+        elif evening_date is not None:
+            if longitude is None:
+                raise ValueError("If evening_date is set, longitude is needed as well")
+            indx = self.index_for_local_evening(evening_date, longitude)
+
         return self.sunsets[indx]
 
     def mjd_indx(self, mjd):
         indx = np.searchsorted(self.sunsets["sunset"], mjd, side="right") - 1
         return indx
 
+    def index_for_local_evening(self, evening_date, longitude):
+        """The index of the night with sunset at a given local date.
+
+        Parameters
+        ----------
+        evening_date : `str` or `datetime.date`
+            The local date of the evening of the night whose index is desired,
+            in ISO8601 format (YYYY-MM-DD).
+        longitude : `float` or  `astropy.coordinates.angles.core.Angle`
+            If a float, then the value is interpreted as being in radians.
+
+        Returns
+        -------
+        night_index : `int`
+            The index of the requested night.
+        """
+        try:
+            longitude = longitude.radian
+        except AttributeError:
+            pass
+
+        if isinstance(evening_date, str):
+            evening_date = datetime.date.fromisoformat(evening_date)
+
+        evening_datetime = datetime.datetime(evening_date.year, evening_date.month, evening_date.day)
+        evening_mjd = np.floor(evening_datetime.timestamp() / (24 * 60 * 60) + 40587)
+
+        # Depending on the time of year, the UTC date rollover might not
+        # always be on the same side of local sunset. Shift by the longitude
+        # to make sure the rollover is always near midnight, far from sunset.
+        matching_nights = np.argwhere(
+            np.floor(self.sunsets["sunset"] + longitude / (2 * np.pi)) == evening_mjd
+        )
+        if len(matching_nights) < 1:
+            raise ValueError(f"Requested night {evening_date} outside of almanac date range")
+
+        night_index = matching_nights.item()
+
+        return night_index
```

Comment:
It would be helpful to know a bit more about where this is being called from. It seems like it's the same (at the end of the day) as mjd_index, but with the added flexility of letting you choose any longitude (any site).  I'm not sure I understand allowing variable longitude, since the Almanac is only created for the Rubin site. 
mjd_index *should* be returning the same thing as index_for_local_evening, if you convert local evening time into MJD UTC, but I can also see how the index_for_local_evening lets you do things that don't make sense (like using a longitude that doesn't match our almanac), and the utility of that isn't clear to me. 
Also, since rubin_scheduler only uses index_mjd internally, it's not obvious to me what this does and how it works with scheduler output. 
So I'm confused.

Reply:
mjd_index "rolls over" at midnight UTC, which sometimes happens during twilight, in which case there can be a few visits at the start of the night where mjd_index and index_for_local_evening can be different. Using index_for_local_evening just makes sure "nights" correspond to actual local nights, rather than UT dates.
Actually using the site longitude was mostly just pedantry on my part. I could just as easily have subtracted an hour or two, and it would have been sufficient.

Reply:
How does this relate to the observing concept of DAYOBS which is 12 hours offset from UTC. Should your concept of night and the summit's concept of night share the same definition?

Reply:
I'm not familiar with DAYOBS, but a simple shift of 12 hours results in a rollover that is always just after sunrise (but sometimes very close to it), so the entire night (sunset to sunrise) will be on the same date. In this context, we're only talking about dates (YYYY-MM-DD or integer MJDs), not date-times, and as long as the whole night (sunset to sunrise) has the same date, I think it's equivalent.

The only ambiguity is the direction of the shift: is the date of the night the date of the evening or of the morning? We've been using the date of the evening, but I've seen other observatories use date of morning, so there's the possibility of an off-by-one problem. I did ask around when I first started working on Rubin observing stuff, and was told that, like other NOAO (at the time) observatories, it uses local date of evening for the date of the night, but my discussion was never at formality level of "this is how DAYOBS is defined", or even that it was called DAYOBS.

There are other contexts where a difference might be important (e.g. assigning visits taken when the sun is up to dates), but that can't ever be relevant to this class,.

Reply:
From SITCOMTN-032:

> dayObs A natural number representing the observation day (in timezone UTC-12:00) 

So, it does look like the earlier (evening) date is the one used, so it should always match.

Reply:
I think that dayobs is probably useful to know about, but not directly useful for our particular use-case ... my understanding is that we're using a datetime.date value instead of MJD primarily for the date-picker for schedview, and inside schedview converting between datetime.date (for the date picker widget) and astropy.time  (for MJD calculations and simulations). rubin_scheduler should probably be primarily concerned with astropy.time and MJD values, although it's possible for simulating an upcoming night that dayobs might be useful (and here I say dayobs just so that it is clear it matches up with summit and DM expectations). And maybe schedview knowing about dayobs would be good, just to help make it clear that we're all talking about the same thing again. 
Schedview could use dayobs and recast into datetime.date if the picker widget is still useful maybe?